File: nucleoid/lang/estree/parser.py
"""
ESTree-compatible parser for JavaScript-like syntax in Python.

This module provides parsing capabilities for JavaScript syntax within
the Nucleoid runtime, converting it to Python-compatible AST structures.
"""

import re
from dataclasses import dataclass
from typing import Any


@dataclass
class ParsedNode:
    """Base class for parsed AST nodes."""
    type: str


@dataclass
class ESTreeNode:
    """ESTree-compatible node structure."""
    type: str
    body: list[Any]
    expression: Any | None = None
    start: int | None = None
    end: int | None = None


class ESTreeParser:
    """Parser for JavaScript-like syntax in Nucleoid."""

    def __init__(self):
        self.node_handlers = {
            'var': self._handle_variable_declaration,
            'let': self._handle_variable_declaration,
            'const': self._handle_variable_declaration,
            'class': self._handle_class_declaration,
            'function': self._handle_function_declaration,
            'if': self._handle_if_statement,
            'for': self._handle_for_statement,
            'return': self._handle_return_statement,
            'throw': self._handle_throw_statement,
            'delete': self._handle_delete_expression,
        }

    def parse(self, code: str, map_nodes: bool = True) -> list[Any]:
        """
        Parse JavaScript-like code into AST nodes.
        
        Args:
            code: The code string to parse
            map_nodes: Whether to map nodes to Nucleoid structures
            
        Returns:
            List of parsed AST nodes
        """
        # For now, this is a simplified implementation
        # In a full implementation, we would use a proper JavaScript parser
        # or implement a complete grammar parser

        # Simple statement splitting and basic parsing
        statements = self._split_statements(code)
        parsed_nodes = []

        for statement in statements:
            try:
                node = self._parse_statement(statement.strip())
                if node:
                    parsed_nodes.append(node)
            except Exception:
                # For now, treat unparseable statements as expressions
                parsed_nodes.append(self._create_expression_node(statement))

        return parsed_nodes

    def parse_function(self, code: str) -> Any:
        """Parse a function string."""
        nodes = self.parse(code, False)
        return nodes[0] if nodes else None

    def _split_statements(self, code: str) -> list[str]:
        """Split code into individual statements."""
        # Simple statement splitting by semicolons and newlines
        # This is very basic - a full implementation would need proper parsing
        statements = []
        current = ""

        for line in code.split('\n'):
            line = line.strip()
            if not line:
                continue

            current += line

            # Simple heuristic for statement completion
            if line.endswith(';') or line.endswith('}') or self._is_complete_statement(line):
                statements.append(current)
                current = ""
            else:
                current += " "

        if current.strip():
            statements.append(current)

        return statements

    def _is_complete_statement(self, line: str) -> bool:
        """Check if a line represents a complete statement."""
        # Simple patterns for complete statements
        patterns = [
            r'^\s*(var|let|const)\s+\w+\s*=.*',
            r'^\s*function\s+\w+\s*\(.*\)\s*{.*}',
            r'^\s*class\s+\w+\s*{.*}',
            r'^\s*if\s*\(.*\)\s*{.*}',
            r'^\s*return\s+.*',
            r'^\s*throw\s+.*',
        ]

        return any(re.match(pattern, line) for pattern in patterns)

    def _parse_statement(self, statement: str) -> Any | None:
        """Parse a single statement."""
        statement = statement.strip().rstrip(';')

        # Variable declaration
        if re.match(r'^\s*(var|let|const)\s+', statement):
            return self._handle_variable_declaration(statement)

        # Assignment expression
        if '=' in statement and not statement.startswith('if') and not statement.startswith('for'):
            return self._handle_assignment_expression(statement)

        # Function declaration
        if statement.startswith('function'):
            return self._handle_function_declaration(statement)

        # Class declaration
        if statement.startswith('class'):
            return self._handle_class_declaration(statement)

        # Control flow statements
        if statement.startswith('if'):
            return self._handle_if_statement(statement)

        if statement.startswith('for'):
            return self._handle_for_statement(statement)

        if statement.startswith('return'):
            return self._handle_return_statement(statement)

        if statement.startswith('throw'):
            return self._handle_throw_statement(statement)

        # Delete expression
        if statement.startswith('delete'):
            return self._handle_delete_expression(statement)

        # Default to expression
        return self._create_expression_node(statement)

    def _handle_variable_declaration(self, statement: str) -> dict[str, Any]:
        """Handle variable declarations (var, let, const)."""
        # Parse: var x = value; or let x = value; or const x = value;
        match = re.match(r'^\s*(var|let|const)\s+(\w+)\s*=\s*(.+)', statement)
        if match:
            kind, identifier, value = match.groups()
            return {
                'type': 'VariableDeclaration',
                'kind': kind,
                'declarations': [{
                    'id': {'type': 'Identifier', 'name': identifier},
                    'init': self._parse_expression(value)
                }]
            }
        return self._create_expression_node(statement)

    def _handle_assignment_expression(self, statement: str) -> dict[str, Any]:
        """Handle assignment expressions."""
        parts = statement.split('=', 1)
        if len(parts) == 2:
            left, right = parts
            return {
                'type': 'AssignmentExpression',
                'left': self._parse_expression(left.strip()),
                'right': self._parse_expression(right.strip())
            }
        return self._create_expression_node(statement)

    def _handle_function_declaration(self, statement: str) -> dict[str, Any]:
        """Handle function declarations."""
        # Very basic function parsing
        return {
            'type': 'FunctionDeclaration',
            'id': {'type': 'Identifier', 'name': 'parsedFunction'},
            'params': [],
            'body': {'type': 'BlockStatement', 'body': []}
        }

    def _handle_class_declaration(self, statement: str) -> dict[str, Any]:
        """Handle class declarations."""
        return {
            'type': 'ClassDeclaration',
            'id': {'type': 'Identifier', 'name': 'ParsedClass'},
            'body': {'body': []}
        }

    def _handle_if_statement(self, statement: str) -> dict[str, Any]:
        """Handle if statements."""
        return {
            'type': 'IfStatement',
            'test': {'type': 'Literal', 'value': True},
            'consequent': {'type': 'BlockStatement', 'body': []},
            'alternate': None
        }

    def _handle_for_statement(self, statement: str) -> dict[str, Any]:
        """Handle for statements."""
        return {
            'type': 'ForOfStatement',
            'left': {'type': 'Identifier', 'name': 'item'},
            'right': {'type': 'Identifier', 'name': 'iterable'},
            'body': {'body': []}
        }

    def _handle_return_statement(self, statement: str) -> dict[str, Any]:
        """Handle return statements."""
        # Extract return value
        value_part = statement[6:].strip()  # Remove 'return'
        return {
            'type': 'ReturnStatement',
            'argument': self._parse_expression(value_part) if value_part else None
        }

    def _handle_throw_statement(self, statement: str) -> dict[str, Any]:
        """Handle throw statements."""
        value_part = statement[5:].strip()  # Remove 'throw'
        return {
            'type': 'ThrowStatement',
            'argument': self._parse_expression(value_part)
        }

    def _handle_delete_expression(self, statement: str) -> dict[str, Any]:
        """Handle delete expressions."""
        target = statement[6:].strip()  # Remove 'delete'
        return {
            'type': 'ExpressionStatement',
            'expression': {
                'type': 'UnaryExpression',
                'operator': 'delete',
                'argument': self._parse_expression(target)
            }
        }

    def _parse_expression(self, expr: str) -> dict[str, Any]:
        """Parse an expression string."""
        expr = expr.strip()

        # Number literal
        try:
            if '.' in expr:
                float(expr)
                return {'type': 'Literal', 'value': float(expr)}
            else:
                int(expr)
                return {'type': 'Literal', 'value': int(expr)}
        except ValueError:
            pass

        # String literal
        if (expr.startswith('"') and expr.endswith('"')) or (expr.startswith("'") and expr.endswith("'")):
            return {'type': 'Literal', 'value': expr[1:-1]}

        # Boolean literals
        if expr == 'true':
            return {'type': 'Literal', 'value': True}
        if expr == 'false':
            return {'type': 'Literal', 'value': False}
        if expr == 'null':
            return {'type': 'Literal', 'value': None}

        # Identifier
        if re.match(r'^\w+$', expr):
            return {'type': 'Identifier', 'name': expr}

        # Default to complex expression
        return {'type': 'Identifier', 'name': expr}

    def _create_expression_node(self, expression: str) -> dict[str, Any]:
        """Create an expression statement node."""
        return {
            'type': 'ExpressionStatement',
            'expression': self._parse_expression(expression)
        }

File: typescript/src/lang/estree/parser.ts
import * as acorn from "acorn";

import $ASSIGNMENT from "../$nuc/$ASSIGNMENT";
import $BLOCK from "../$nuc/$BLOCK";
import $CLASS from "../$nuc/$CLASS";
import $DELETE from "../$nuc/$DELETE";
import $EXPRESSION from "../$nuc/$EXPRESSION";
import $FOR from "../$nuc/$FOR";
import $FUNCTION from "../$nuc/$FUNCTION";
import $IF from "../$nuc/$IF";
import $INSTANCE from "../$nuc/$INSTANCE";
import $RETURN from "../$nuc/$RETURN";
import $THROW from "../$nuc/$THROW";
import { Identifier } from "acorn";

interface Node {
  type: string;
  [key: string]: any;
}

interface EstreeNode {
  type: string;
  body: any[];
  expression?: any;
  start?: number;
  end?: number;
  [key: string]: any;
}

interface DeclarationNode extends Node {
  kind: string;
  declarations: {
    id: any;
    init: any;
  }[];
}

interface ClassNode extends Node {
  id: any;
  body: {
    body: any[];
  };
}

interface AssignmentNode extends Node {
  left: any;
  right: any;
}

interface BlockNode extends Node {
  body: any[];
}

interface IfNode extends Node {
  test: any;
  consequent: any;
  alternate: any;
}

interface ExpressionNode extends Node {
  expression: {
    type: string;
    operator?: string;
    argument?: any;
  };
}

interface NewNode extends Node {
  callee: any;
  arguments: any[];
}

interface ReturnNode extends Node {
  argument: any;
}

interface MethodNode extends Node {
  key: any;
  value: {
    params: any[];
    body: any;
  };
}

interface FunctionNode extends Node {
  id: any;
  params: any[];
  body: any;
}


interface ForOfNode extends Node {
  left: Identifier;
  right: any;
  body: {
    body: any[];
  };
}

interface ThrowNode extends Node {
  argument: any;
}

function parse(string: string, map: boolean = true) {
  const estree = acorn.parse(string, {
    ecmaVersion: 2020,
    allowReturnOutsideFunction: true,
  }) as EstreeNode;

  removeLocations(estree);

  return map
    ? estree.body.map(parseNode).filter((exp) => exp)
    : estree.body[0].expression;
}

function parseFunction(string: string) {
  const estree = acorn.parse(string, {
    ecmaVersion: 2020,
    allowReturnOutsideFunction: true,
  }) as EstreeNode;

  if (estree.body[0].type === "ExpressionStatement") {
    return estree.body[0].expression;
  } else {
    return estree.body[0];
  }
}

function parseNode(node: Node) {
  switch (node.type) {
    case "VariableDeclaration": {
      const {
        kind,
        declarations: [declaration],
      } = node as DeclarationNode;

      const { id, init } = declaration;

      return $ASSIGNMENT(
        kind.toUpperCase() as "VAR" | "LET" | "CONST",
        id,
        init
      );
    }
    case "ClassDeclaration": {
      const {
        id,
        body: { body },
      } = node as ClassNode;

      const methods = body.map((method) => parseNode(method));

      return $CLASS(id, methods);
    }
    case "AssignmentExpression": {
      const { left, right } = node as AssignmentNode;
      return $ASSIGNMENT(null as null, left, right);
    }
    case "BlockStatement": {
      const statements = (node as BlockNode).body.map(parseNode);
      return $BLOCK(statements);
    }
    case "IfStatement": {
      const { test, consequent, alternate } = node as IfNode;
      const parsedConsequent = parseNode(consequent);
      const parsedAlternate = alternate ? parseNode(alternate) : undefined;

      return $IF(test, parsedConsequent, parsedAlternate);
    }
    case "ExpressionStatement": {
      const { expression } = node as ExpressionNode;

      if (["AssignmentExpression", "NewExpression"].includes(expression.type)) {
        return parseNode(expression);
      } else if (
        expression.type === "UnaryExpression" &&
        expression.operator === "delete"
      ) {
        return $DELETE(expression.argument);
      } else {
        return $EXPRESSION(expression);
      }
    }
    case "NewExpression": {
      const { callee, arguments: args } = node as NewNode;
      return $INSTANCE(callee, undefined, undefined, args[0]);
    }
    case "ReturnStatement": {
      const { argument } = node as ReturnNode;
      const statements = parseNode(argument);
      return $RETURN(statements);
    }
    case "MethodDefinition": {
      const { key, value } = node as MethodNode;

      return $FUNCTION(key, value.params, parseNode(value.body));
    }
    case "FunctionDeclaration": {
      const { id, params, body } = node as FunctionNode;

      return $FUNCTION(id, params, parseNode(body));
    }
    case "ForOfStatement": {
      const { left, right, body } = node as ForOfNode;

      const bodyExpressions = body.body.map(parseNode);
      return $FOR(left as Identifier, right, bodyExpressions);
    }
    case "ThrowStatement": {
      const { argument } = node as ThrowNode;

      return $THROW(argument);
    }
    case "EmptyStatement": {
      return;
    }
    default: {
      return $EXPRESSION(node);
    }
  }
}

function removeLocations(estree: any): void {
  for (let key in estree) {
    delete estree.start;
    delete estree.end;

    if (estree[key] !== null && typeof estree[key] === "object") {
      removeLocations(estree[key]);
    } else if (Array.isArray(estree[key])) {
      estree[key].forEach((node: any) => removeLocations(node));
    }
  }
}

export { parse, parseFunction };
